<?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />
  <arg name="ip" default="192.168.0.20" />
  <arg name="port" default="3883" />
  <arg name="uri" default="radio://0/110/2M/0xE7E7E7E704" />
  <arg name="frame" default="crazyflie" />
  

  <arg name="mpc_command_subscribe_topic_name" default="/command/roll_pitch_yawrate_thrust" />
  <arg name="publish_topic_name" default="cmd_vel" />
  <arg name="ThrustRateConstant" default="0.000943" />
  <arg name="ThrustOffsetConstant" default="-5.537574" />
  <arg name="Newton2GramForceConstant" default="0.0098066" />
  <arg name="MPC_Frequency" default="50.0" />
  <arg name="uav_pose_reference_topic_name" default="/command/pose" />
  <arg name="takeoff_height" default="1.0" />

<!--  This is supposed to record bag files to record from launch, but couldn't get it to work :(
  <arg name="record_path" default="~/Desktop" />
  <arg name="bag_name" default="sup_then" />
  <arg name="record_topic1" default="/crazyflie/odom" />
  <arg name="record_topic2" default="/crazyflie/cmd_vel" />
  <arg name="record_topic3" default="/crazyflie/imu" />
-->

  <!-- initialize server -->
  <include file="$(find crazyflie_driver)/launch/crazyflie_server.launch">
  </include>

  <node name="joy" pkg="joy" type="joy_node" output="screen">
     <param name="dev" value="$(arg joy_dev)" />
  </node>

  <group ns="crazyflie">
    <include file="$(find crazyflie_driver)/launch/crazyflie_add.launch">
      <arg name="uri" value="$(arg uri)" />
      <arg name="tf_prefix" value="crazyflie" />
      <arg name="enable_logging" value="True" />
    </include>
  
<!--
    <include file="$(find crazyflie_controller)/launch/crazyflie2.launch">
      <arg name="frame" value="$(arg frame)" />
    </include>
-->
    <node name="CrazyflieWrapperNode" pkg="flyboi_ethz" type="CrazyFlieWrapperNode.py" output="screen">
      <!-- Params for CrazyFlieSubscribeThustClass -->
      <param name="rate" value="$(arg MPC_Frequency)" />
      <param name="mpc_command_subscribe_topic_name" value="$(arg mpc_command_subscribe_topic_name)" />
      <param name="publish_topic_name" value="$(arg publish_topic_name)" />
      <param name="ThrustRateConstant" value="$(arg ThrustRateConstant)" />
      <param name="ThrustOffsetConstant" value="$(arg ThrustOffsetConstant)" />
      <param name="Newton2GramForceConstant" value="$(arg Newton2GramForceConstant)" />
      <!-- Params for Pose Reference -->
      <param name="uav_pose_reference_topic_name" value="$(arg uav_pose_reference_topic_name)" />
      <param name="takeoff_height" value="$(arg takeoff_height)" />
    </node>

    <node name="PID_Controller" pkg="flyboi_ethz" type="Crazyflie_PID_Node" output="screen">
	<rosparam command='load' file="$(find flyboi_ethz)/resources/PID/crazyflie.yaml"/>
    </node>

    <node pkg="tf" type="static_transform_publisher" name="baselink_broadcaster" args="0 0 0 0 0 0 1 $(arg frame) /crazyflie/base_link 100" />

    <!-- run vrpn client -->
    <node pkg="vrpn_client_ros" name="vrpn_client_node" output="screen" type="vrpn_client_node" required="true">
      <param name="server" value="$(arg ip)" />
      <param name="port" value="$(arg port)" />
      <param name="update_frequency" value = "200" />
      <param name="frame_id" value = "/world" />
      <param name="use_server_time" value = "false" />
      <param name="broadcast_tf" value = "true" />
      <param name="refresh_tracker_frequency" value = "1.0" />
    </node>
    
    <node name="publish_external_position_vrpn" pkg="crazyflie_demo" type="publish_external_position_vrpn.py" output="screen" />
    
    <node name="Joystick_Control" pkg="flyboi_ethz" type="Joystick_Control_Node.py" output="screen">
      <param name="joy_topic" value="/joy"/>
    </node>

  </group>

<!-- Record Some Data
 
    <node name="rosbag" pkg="rosbag" type="record" output="screen" respawn = "true" respawn_delay = "5"
      arg="-O test1.bag /crazyflie/odom /crazyflie/cmd_vel /crazyflie/imu" />
 -->
  

 <!--
 run rviz 

  <param name="robot_description" command="$(find xacro)/xacro.py $(find crazyflie_description)/urdf/crazyflie.urdf.xacro" />
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyflie_demo)/launch/crazyflie_pos.rviz" required="true" />
-->
</launch>
